Adding helm chart to CI

donnyyung opened the pull request:
This PR does a few things:

- Makes the helm chart lintable
- Adds that lint to our CI process
- Keeps the helm chart in sync with the latest GA version of telepresence.
- Adds a script for publishing the helm chart
- Creates CI for publishing helm chart versions.

Here are the proposed changes to the release process, additions are in green: https://www.notion.so/datawire/Proposal-with-Helm-Chart-To-Release-Telepresence-2-x-x-130753a3fa7a4691bee3d47b767ff64e

I have ran the script for publishing to the `charts-dev` path without trouble, so I believe I have tested it to the best of my ability. Obviously if the first production release does not work, I will debug and get it fixed. 

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes. -- todo
 - [x] My change is adequately tested.
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

Commit:
Make helm lint work

The chart does some validation around the .Release.Name but since we
can't set the .Release.Name in `helm lint`, we had to add a hidden value
to ignore that validation.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add chart linting to CI

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add chart publishing to CI

This adds the push_chart.sh script which contains the logic for
packaging and pushing the helm chart to the S3 bucket, as well as the
circle changes required to do this on specified tags.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Automatically bump appVersion for chart

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Updating default + prod paths for chart publishing

Earlier I was using a test path for inside of the AWS Bucket. Alix has
since set up the chart repositories and the jobs generating metadata
once charts are pushed, so I've updated the script + circle config to
use the correct values

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add comment linking to rbac docs in chart

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
bump changelog

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Make chart work with 2.3.1

2.3.1 added some new yaml files required for the agentInjectorService,
so I've updated the helm chart to also support those things.  Also added
some additional templating.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

LukeShu commented on build/go.mk:
```diff
@@ -107,6 +107,10 @@ lint: lint-deps ## (Lint) Run the linters (golangci-lint and protolint)
+.PHONY: lint-chart
+lint-chart:
+	helm lint charts/telepresence --set isCI=true
+
```
Is there a good reason this isn't part of the main `make lint`?

donnyyung replied:
Not sure if a _good_ reason, but in case somebody wanted to _only_ lint the chart or _only_ do the go lint, having them separate would be easier... but the commands themselves aren't that complicated so maybe it's cleaner to combine them and then someone can look the commands up if they only want to do one or the other.

LukeShu commented on .circleci/config.yml:
```diff
@@ -160,6 +177,18 @@ jobs:
+    # we have a need for rc charts), but explicitly calling them out here
+    # in case we change our mind
+    - run: |
+        [[ $CIRCLE_TAG == "*rc*" ]] || BUCKET_DIR="charts" packaging/push_chart.sh
```
1. Drop the quotes around `"*rc*"`; because of the double quotes, it'll match `*rc*` literally rather than as a glob.  If this is surprising to you, then it may be helpful to know that despite the visual similarity to the `[` command, the `[[ … ]]` construct is a special piece of Bash syntax, rather than a regular command.
2. I'd be more comfortable excluding _all_ prerelease tags, by having it by `[[ $CIRCLE_TAG == *-* ]]`.

LukeShu replied:
Actually: See my next comment below.

LukeShu replied:
The `[[ … ]]` check is still broken.

LukeShu commented on .circleci/config.yml:
```diff
@@ -191,6 +220,13 @@ workflows:
+          context: telepresence2-release
+          filters:
+            tags:
+              only: /^chart-v.*/
```
1. Do we expect to be doing chart releases that don't coincide with a Telepresence release?  IMO, this would be a good opportunity to version them in lockstep.
2. It would be good to just modify the regex here instead of having the `[[ $CIRCLE_TAG == … ]]` check in the job.  I believe `/^chart-v(.(?!rc))*$/` would do what you want (the Circle [docs](https://circleci.com/docs/2.0/workflows/#using-regular-expressions-to-filter-tags-and-branches) say that it uses [`java.util.regex` syntax](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html)).
3. So using my suggestion of excluding all prerelease tags, the regex would be `/^chart-v[^-]*$/`

donnyyung replied:
1. So in an ideal world, no. and I had originally planned to keep the versions in-sync. But after chatting with people on edge stack, they noted the possibility that sometimes you might need to only change the chart (e.g. an error in the yaml when a user uses some value, which isn't used in our smoke tests).  Which maybe for edge stack that is a problem since they release less frequently, but we could do another release shortly after the problem was discovered "x.y.z+1" with a changelog that's like, "Bugfix: Fixes issue with helm chart".  So if we aren't opposed to doing something like that, then I think for our own sanity it makes sense to keep the versions in lockstep.

2/3. Good idea, will do if ^ isn't a good workflow and we can't keep them in lockstep.

LukeShu replied:
1. Given the goal of keeping telepresence releases light, I have no problem "Bugfix: Helm chart" releases.  So in my mind, that means not having separate `chart-v*` tags.

donnyyung replied:
Sounds good, I will remove the `chart-v*` and fold helm chart publish into our already existing release steps then.

Commit:
Enable installing tm via chart for smoke tests

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Helm chart support 2.3.2

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Changes for PR comments

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Reorganized rbac files for chart

rbac used to just apply to (non-privileged) users. But now that we have
Traffic Manager rbac, I had to update the values file to account for the
two categories of rbac that we have.  Additionally, I condensed all the
trafficManager rbac files into one (trafficManagerRbac) to keep it
organized and match clientRbac (formerly rbac.yaml)

Signed-off-by: Donny Yung <donaldyung@datawire.io>

LukeShu commented on .circleci/config.yml:
```diff
@@ -160,6 +176,18 @@ jobs:
+    steps:
+    - checkout
+    - install-helm
+    - run: make lint
```
Why have a `make lint` here? We don't do any other checks in the release jobs.

donnyyung replied:
I was being too defensive since I wanted to ensure the chart linted prior to pushing, but since we ensure lint on merge to master, then there's really no _real_ reason to lint again. So I'll remove it.

LukeShu commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ### 2.3.3 (TBD)
 
 - Feature: `telepresence intercept` now supports a `--to-pod` flag that can be used to port-forward sidecars' ports from an intercepted pod
+- Feature: The helm chart is now published as part of our CI.
```
Nit: Capitalize "Helm"

LukeShu commented on charts/telepresence/README.md:
```diff
@@ -27,30 +27,45 @@ The following tables lists the configurable parameters of the Ambassador chart a
 | image.pullPolicy         | How the `Pod` will attempt to pull the image.                                                                           | `IfNotPresent`                                                                                    |
 | image.tag                | Override the version of the Traffic Manager to be installed.                                                            | `""` (Defined in `appVersion` Chart.yaml)                                                         |
 | image.imagePullSecrets   | The `Secret` storing any credentials needed to access the image in a private registry.                                  | `[]`                                                                                              |
+| ports            | The ports the Traffic Manager `Deployment` will expose. **Do not change.**                               | `[{"name":"api","port":8081},{"name": "https", "port": 8443]` |
```
What's the point in having a setting if we tell people not to set it?

donnyyung replied:
I saw that this is what we did with service.ports, but I think it doesn't really make sense / is kinda confusing, so removing the "do not changes" I added, as well as the pre-existing ones and will have it hard-coded in the yaml.

LukeShu commented on packaging/push_chart.sh:
```diff
@@ -0,0 +1,76 @@
+    bucket_dir="charts-dev"
+fi
+
+if [ -z "$AWS_BUCKET"] ; then
```
You definitely need a space between the `"` and the `]`.

On that note, running `shellcheck` on our shell scripts from `make lint` would be a good enhancement.

LukeShu commented on smoke-tests/run_smoke_test.sh:
```diff
@@ -437,6 +471,13 @@ finish_step
 ##########################################################
 
+# First we uninstall the helm chart if it was used
+if [ ! -z $USE_CHART ]; then
```
Either use double `[[ brackets ]]` or put `"quotes"` around the variable.

LukeShu commented on .circleci/config.yml:
```diff
@@ -59,6 +59,21 @@ commands:
+            chmod 700 get_helm.sh
+            ./get_helm.sh --version << parameters.version >>
+            rm ./get_helm.sh
+
```
Since we're relying on `helm lint` in `make lint`, which is something we expect developers to run, I think I'd prefer to have this in `tools.mk`

Commit:
More PR feedback

This includes various changes:
Installing helm in build-aux/tools.mk
Removing installation of helm in circle bc of ^
Linting packaging/push_chart.sh
Removed service.ports since it can't be changed

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Version helm chart w/ telepresence version

Signed-off-by: Donny Yung <donaldyung@datawire.io>
